Add Interstitial Tag Support

rmigneco opened the pull request:
### Description

This PR adds support for Interstitial Tag Attributes as outlined in the [Appendix D of the HLS Spec](https://datatracker.ietf.org/doc/html/draft-pantos-hls-rfc8216bis#appendix-D)

### Change Notes

* Adds `PantosValue` attributes to support interstitials
* Updates the `EXT_X_DATERANGETagValidator` to include additional values and include specific validation rules for HLS Interstitials
* Adds unit tests for the new validation rules
* Creates `HLSInterstitialValues` to represent the attribute options 
* Creates `InterstitialTagBuilder` which facilitates creating a `HLSTag` type configured specifically for interstitials
* Unit tests for `InterstitialTagBuilder`

### Pre-submission Checklist

- [x] I ran the unit tests locally before checking in.
- [x] I made sure there were no compiler warnings before checking in.
- [x] I have written useful documentation for all public code.
- [x] I have written unit tests for this new feature.

rmigneco committed:
Added PantosValue for Interstitial Tags

rmigneco committed:
Added Validation Issues for HLS Interstitial tags

- checking that with HLS CLASS attribute, only asset list or id can be accepted

rmigneco committed:
Added Special Value types for interstital attributes

- completed the Generic Tag validator for Date Range

rmigneco committed:
Added additional date range validation tests

rmigneco committed:
Added Interstitial Value Tests

- testing seek restrictions and alignment
- renamed Interstitial Alignment

rmigneco committed:
Added interstitial tag builder

- included required initializers
- included builder functions

rmigneco committed:
Changed visibility of Dateformatter so they could be re-use

rmigneco committed:
Added discardable result to builder functions

- completed tag return in build func
- added documentation

rmigneco committed:
Added Tag Builder Tests

- ensure HLS Interstitial Value types conform to equatable

rmigneco committed:
Removed TODO comment

theRealRobG commented on mambaSharedFramework/HLSInterstitialValueTypes.swift:
```diff
@@ -0,0 +1,112 @@
+public struct HLSInterstitialAlignment: FailableStringLiteralConvertible, Equatable {
+
+    public enum Snap: String, CaseIterable {
+        /// client SHOULD locate the segment boundary closest to the scheduled resumption point from the
+        /// interstitial in the Media Playlist of the primary content and resume playback of primary content at that boundary.
+        case snapIn = "IN"
+        
+        /// client SHOULD locate the segment boundary closest to the START-DATE of the interstitial in the
+        /// Media Playlist of the primary content and transition to the interstitial at that boundary.
+        case snapOut = "OUT"
```
Total nit-pick, but `snap` is already in the enum name, so feels like `.in` and `.out` are better case names (as well as being more terse):
```suggestion
        /// client SHOULD locate the segment boundary closest to the scheduled resumption point from the
        /// interstitial in the Media Playlist of the primary content and resume playback of primary content at that boundary.
        case in = "IN"
        
        /// client SHOULD locate the segment boundary closest to the START-DATE of the interstitial in the
        /// Media Playlist of the primary content and transition to the interstitial at that boundary.
        case out = "OUT"
```
But really, I don't mind that much... Just calling it out in case it was an oversight.

rmigneco replied:
No that's good feedback, will update

theRealRobG commented on mambaSharedFramework/InterstitialTagBuilder.swift:
```diff
@@ -0,0 +1,293 @@
+
+import Foundation
+
+
```
Another nit-pick (whitespace):
```suggestion
```

theRealRobG commented on mambaSharedFramework/InterstitialTagBuilder.swift:
```diff
@@ -0,0 +1,293 @@
+    let assetList: String?
+    
+    /// the duration of the interstitial content in seconds
+    var duration: Double?
```
It is also useful to include `plannedDuration`. This helps AVPlayer know what range an interstitial _could_ occupy when the daterange `DURATION` is unknown without having to resolve the asset list/uri. This is particularly useful for ad insertions on the live edge, where the `PLANNED-DURATION` for the break is known, but we don't want to set the `DURATION` yet because there may be an early return to action where the break would need to be cut short.

```suggestion
    var duration: Double?
    
    /// the expected duration of the interstitial content in seconds which can indicate a value when the actual duration is not yet known
    var plannedDuration: Double?
```

theRealRobG commented on mambaSharedFramework/InterstitialTagBuilder.swift:
```diff
@@ -0,0 +1,293 @@
+        }
+        
+        if let duration {
+            hlsTagDictionary[PantosValue.duration.rawValue] = HLSValueData(value: String(duration), quoteEscaped: true)
```
Should this be `quoteEscaped`? Doesn't this imply the attribute gets created like `DURATION="60.0"` (for example)?

```suggestion
            hlsTagDictionary[PantosValue.duration.rawValue] = HLSValueData(value: String(duration), quoteEscaped: false)
```

Same question for `resumeOffset` and `playoutLimit` which are defined as "signed-decimal-floating-point" and "decimal-floating-point" respectively.

rmigneco replied:
yeah this was a miss. I've updated it to use `false` for those decimal values.

theRealRobG reviewed (commented):
Looks good! Just some small comments and questions.

rmigneco committed:
Updated Snap enum case names

- updated affected tests

rmigneco committed:
Trim whitespace

rmigneco committed:
Added `plannedDuration` to interstitial tag builder

- fixed qutoe escape issues for numerical values in builder
- upated unit tests

rmigneco commented:
> Looks good! Just some small comments and questions.

@theRealRobG got your changes incorporated - LMK if you'd like to get this into 2.0 as well
